### ISSUE TITLE
Unable to find trailer dictionary while recovering damaged file

### ISSUE THREAD

**Issue by FelixFrog** [qpdf/qpdf#573]
Hello everyone! I have a script which parses a set of pdf files (pages of a book to be precise) and stitches them together with pikepdf. Since these files are recovered in a strange way the xref table obviously are mangled up, thus making me rely on qpdf's recovering abilities. But sometimes certain files are unrecoverable, spitting out this error:

```
WARNING: 0203.pdf: file is damaged
WARNING: 0203.pdf (offset 102607): xref not found
WARNING: 0203.pdf: Attempting to reconstruct cross-reference table
0203.pdf: unable to find trailer dictionary while recovering damaged file
```

The first three lines are expected, the fourth one is not though, since it's strange that it cannot find the trailer dictionary. The file in question can be opened by almost every pdf viewer I tried without problems. Also analyzing it in a hex editor doesn't seem to reveal any damage that could affect qpdf's recovery.

I'm on qpdf version 10.3.2 on macOS 11.4. The file in question is 0203.pdf, while I attach some other files among the hundreds that gave me no problem.

Thanks in advance

[0203.pdf](https://github.com/qpdf/qpdf/files/7490809/0203.pdf)
[0202.pdf](https://github.com/qpdf/qpdf/files/7490810/0202.pdf)
[0201.pdf](https://github.com/qpdf/qpdf/files/7490811/0201.pdf)
[0196.pdf](https://github.com/qpdf/qpdf/files/7490812/0196.pdf)

**Comment by jberkenbilt** [comment]
Good timing as I'm actively working on some qpdf issues. I am able to reproduce this, and a quick glance at the file makes me think it should be recoverable. I'll study this in depth and see if I can make qpdf able to recover it. If we're lucky, I can get a fix out with 10.4, which I'm hoping to release in the next several days. No promises. :-)

**Comment by jberkenbilt** [comment]
Turns out this was an easy fix, so I will definitely get it in for 10.4.0. The xref recovery code was making some unnecessary assumptions about the file. This should actually help quite a number of cases.

**Comment by jberkenbilt** [comment]
For the record, the issue here was that qpdf was scanning for objects in xref recovery and expecting `endobj` to be at the beginning of the line. Until it found `endobj`, it was not looking for new objects. This means that it doesn't have to tokenize most of the file, but the downside is that it will not handle cases where endobj is not at the beginning of the line. When I originally wrote that code, it used regular expressions to look at each line, which was relatively slow, but years ago, I removed that and rewrote it to use the QPDF tokenizer, which is pretty fast. I removed this restriction, and it is now able to recover not only this file but make better progress on two other files in the qpdf test suite. I suspect this will significantly improve qpdf's recovery ability at the cost of some performance on the recovery operation itself. It will also make qpdf mess up on pathological cases where embedded objects contain text that looks like other objects, but this is probably much less common than the case in this file.

@FelixFrog Thanks for taking the time to report this. I think it will probably make a difference for lots of users.
